### PR TITLE
Feature/broadcast tool

### PR DESCRIPTION
**Pull request by Kiryous**
DEMO:
https://www.loom.com/share/d75c187de63a4592847993791abb0718?from_recorder=1&focus_title=1

TODO: 
- [x] Success state
- [x] Fix auto height

### CODE REVIEW

File: packages/dialect-react-ui/components/Broadcast/BroadcastForm.tsx [hunk 1]
```diff
@@ -0,0 +1,94 @@
...
+  if (isFetchingAddresses) {
+    usersString = 'Loading users...';
+  } else if (noUsers) {
+    usersString = 'No users yet';
+  }
+
+  const sendBroadcastMessage = async () => {
+    setIsSending(true);
+    try {
+      await dapp.messages.send({ title, message });
```
Comment (fsher) [inline packages/dialect-react-ui/components/Broadcast/BroadcastForm.tsx:35]: Should we clear the error when we try and send again for instance?

**Review by fsher — APPROVED** [review]
Minor improvements. LGTM, well done sir!

File: packages/dialect-react-ui/components/Broadcast/BroadcastForm.tsx [hunk 2]
```diff
@@ -0,0 +1,94 @@
...
+function BroadcastForm({ dapp }: BroadcastFormProps) {
+  const { addresses, isFetching: isFetchingAddresses } = useDappAddresses();
+  const { textStyles, colors, outlinedInput, textArea } = useTheme();
+  // Consider moving error handling to the useDapp context
+  const [error, setError] = useState<Error | null>(null);
+  const [isSending, setIsSending] = useState(false);
+  const [title, setTitle] = useState('');
+  const [message, setMessage] = useState('');
+
+  const usersCount = addresses?.length || 0;
```
Comment (tsmbl) [inline packages/dialect-react-ui/components/Broadcast/BroadcastForm.tsx:21]: Dapp address is actually not a user
Users count = set of unique wallets, associated with enabled dapp addresses, associated with verified addresses

File: packages/dialect-react-ui/components/Broadcast/BroadcastForm.tsx [hunk 3]
```diff
@@ -0,0 +1,94 @@
...
+          placeholder="Write message..."
+          onChange={(e) => {
+            setMessage(e.target.value);
+          }}
+          value={message}
+          // FIXME: add the outlined texarea class to the theme
+          className={clsx(outlinedInput, 'dt-w-full dt-h-44')}
+        />
+        <div className="dt-text-xs dt-pl-1 dt-opacity-50">
+          Limit: {message.length}/500
```
Comment (tsmbl) [inline packages/dialect-react-ui/components/Broadcast/BroadcastForm.tsx:72]: Should be length of ut8-encoded representation i.e. TextEncoder.encode().length / 800

File: packages/dialect-react-ui/components/Broadcast/BroadcastForm.tsx [hunk 4]
```diff
@@ -0,0 +1,94 @@
...
+  const { textStyles, colors, outlinedInput, textArea } = useTheme();
+  // Consider moving error handling to the useDapp context
+  const [error, setError] = useState<Error | null>(null);
+  const [isSending, setIsSending] = useState(false);
+  const [title, setTitle] = useState('');
+  const [message, setMessage] = useState('');
+
+  const usersCount = addresses?.length || 0;
+  const noUsers = usersCount === 0;
+  const disabled = !title || !message || noUsers;
```
[the remark this answers is not in the document]
Comment (Kiryous) [reply packages/dialect-react-ui/components/Broadcast/BroadcastForm.tsx:23]: yeap!

**Review by tsmbl — CHANGES_REQUESTED** [review]
Overall cool, adding some comments on sdk usage

File: packages/dialect-react-ui/components/Broadcast/BroadcastForm.tsx (hunk 3 above)
Comment (Kiryous) [reply packages/dialect-react-ui/components/Broadcast/BroadcastForm.tsx:72]: Do we also need to add title before calculating?

Comment (tsmbl) [reply packages/dialect-react-ui/components/Broadcast/BroadcastForm.tsx:72]: No, just text

Comment (Kiryous) [reply packages/dialect-react-ui/components/Broadcast/BroadcastForm.tsx:72]: cool. is the limit is 800 or 500 as designs suggest?

Comment (tsmbl) [reply packages/dialect-react-ui/components/Broadcast/BroadcastForm.tsx:72]: Actually it's 1024 utf8 bytes (soft limit rn), I propose to use it

File: packages/dialect-react-ui/components/Broadcast/BroadcastForm.tsx (hunk 2 above)
Comment (Kiryous) [reply packages/dialect-react-ui/components/Broadcast/BroadcastForm.tsx]: Got it, fixed!

File: packages/dialect-react-sdk/hooks/useDapp.ts [hunk 5]
```diff
@@ -0,0 +1,41 @@
+import type {
+  Dapp,
+  DialectSdkError,
+  DialectWalletAdapter,
+} from '@dialectlabs/sdk';
+import useSWR from 'swr';
+import { EMPTY_OBJ } from '../utils';
+import useDialectSdk from './useDialectSdk';
+
+const DAPP_CACHE_KEY = (wallet: DialectWalletAdapter) =>
```
Comment (zhelezkov) [inline packages/dialect-react-sdk/hooks/useDapp.ts:10]: let's move it to hooks/internal/swrCache.ts

File: packages/dialect-react-sdk/hooks/useDappAddresses.ts [hunk 6]
```diff
@@ -0,0 +1,53 @@
+import type { Dapp, DappAddress, DialectSdkError } from '@dialectlabs/sdk';
+import { useEffect } from 'react';
+import useSWR from 'swr';
+import { EMPTY_ARR, EMPTY_OBJ } from '../utils';
+import useDapp from './useDapp';
+
+const DAPP_ADDRESSES_CACHE_KEY = (dapp: Dapp | null) =>
```
Comment (zhelezkov) [inline packages/dialect-react-sdk/hooks/useDappAddresses.ts:7]: also let's move it to the same place where all other cache keys

**Review by zhelezkov — APPROVED** [review]
lgtm. minor q. approving in advance

**Review by tsmbl — APPROVED** [review]
Nice job, lgtm

**Pull request merged**